doc/source/using_config.rst

```diff
@@ -178,6 +189,19 @@ Attributes
   * ``auto``: Only cache the build trees where necessary (e.g. for failed builds)
   * ``always``: Always cache the build tree.
 
+* ``storage-service``
+
+  An optional :ref:`service configuration <user_config_remote_execution_service>`
+  to use a *Content Addressable Storage* service as a remote cache. Write access
+  is required.
+
+  This allows operation with a small local cache even with large projects.
+  This is especially useful in combination with
+  :ref:`remote execution <user_config_remote_execution>` to minimize downloads
+  of build outputs, which may not be needed locally.
+
+  CAS objects will be downloaded on demand, e.g. on ``bst artifact checkout``.
```

Comment:
What is the behavior of `bst pull` here ?

I think it's important that users have the ability to pull artifacts with the confidence that they will exist in their local CAS after successfully pulling (e.g. for offline use in the near future).

In fact, I am hard pressed to imagine a use case where the user wants to explicitly invoke `bst pull` but not want the artifacts to be on their client machine (although I would not have this expectation for the cases where `pull` is an implied side effect of something else, like in the case of implied pull in `bst push`)

Reply:
The contents of the `storage-service` are considered part of the cache, so blobs are not guaranteed to exist locally with the current branch after `bst artifact pull` (or `bst source fetch`).

It could make sense in some cases. E.g. when the `storage-service` in the cache configuration is on a local network and you're pulling from an artifact server outside the local network. Or if you want to separately call `bst source fetch` before `bst build` with remote execution, you may not need the sources locally, you just want them in the cache `storage-service`, ready to be used for the remote execution build.

However, I can definitely see that this behavior could be confusing, especially if the cache `storage-service` is not on the local network. Technically it is consistent with other commands, though.

I'm not sure whether it would be ideal to simply ensure all blobs are in the local cache after `bst artifact pull` (and `bst source fetch`) as the behavior would be somewhat inconsistent. And with a low local cache quota it might even fail.

A possible alternative would be to print a warning/info message when running `bst artifact pull` with a cache `storage-service` configured and add a global CLI option to disable the cache `storage-service` without having to edit `buildstream.conf`.

I don't know whether it will be common for people to frequently switch between configurations with a cache `storage-service` and offline operations. My guess is that people who use remote execution will not frequently work offline (with the same project).

Any thoughts which solution makes most sense?

Reply:
> I'm not sure whether it would be ideal to simply ensure all blobs are in the local cache after `bst artifact pull` (and `bst source fetch`) as the behavior would be somewhat inconsistent. And with a low local cache quota it might even fail.

My thoughts were that this would only happen if pulling artifacts or sources was *explicitly* requested, i.e. the user explicitly typed `bst artifact pull` or `bst source fetch`, but not in the case that these activities were a side effect of building, where the user has not explicitly asked for sources or artifacts.

The premise of course being that if a user types either of those commands, it is *because* they want the data be available for offline use, and I can't imagine any other reason at the moment why a user would type `bst artifact pull` or `bst source pull`

> A possible alternative would be to print a warning/info message when running `bst artifact pull` with a cache `storage-service` configured and add a global CLI option to disable the cache `storage-service` without having to edit `buildstream.conf`.

I'd be happy with a warning without necessarily adding a CLI option, (if we were to add a CLI option it might be desirable to support a value like we do for `RemoteSpec` with an ability to specify the value as `local`)... I'm more concerned with messaging and clarity of what's going on than I am with flexibility here.

> I don't know whether it will be common for people to frequently switch between configurations with a cache `storage-service` and offline operations. My guess is that people who use remote execution will not frequently work offline (with the same project).

Interesting... first... I think my biggest concern with *"My guess is that people who use remote execution will not frequently work offline"*, is that we have not made it clear to the user that the purpose of this option is *for use with remote execution*, we have only recommended it for such use ~~without articulating it's usefulness in other contexts~~ (looking at the update patch, it seems we are a little bit closer to articulating this).

Orthogonal to the above concern, I don't really share this expectation about how people will use remote execution, I think that:
* People who use remote execution will probably use a `buildstream.conf` that is tightly coupled with each individual BuildStream project they work on on their workstation
* Even for projects which they use remote execution for, they will find themselves in situations where they will want to edit their `buildstream.conf` and download everything they need onto their laptop, so that they can run a few tests on their 14hrs flight on the way to the conference they are going to where the demo they have to prepared really has to work.

### Summary

* Let's make sure there is a warning when the user explicitly tries to pull things and those things are not actually being delivered to the user
* Let's still try to be more explicit in the documentation about `storage-service`, even if it demands the user be a bit more exposed to internals
  * In this case, perhaps we should explain that by default all content is stored locally and that this option causes content to be stored remotely, and separately to this explanation, we can go on to outline how this can be useful with and without remote execution (while explicitly recommending it for remote execution)

### Separate enhancement

Beyond this, logging about pulling and pushing artifacts should be enhanced.

When displaying log lines which express where artifacts and sources are being pushed *to* or pulled *from*, we should explicitly also list the URI of the other side of the endpoint, since it might not be local anymore (while perhaps using the word "localhost" in the cases that `storage-service` is not configured).

This should help keep things clear about what is going on when things are pulled and pushed.

Maybe this is already done in your patch which I didn't review in depth, but if not... then I don't mind postponing this bit and would ask that you open a separate issue to track the logging adjustment.

Reply:
Thanks for your comments.

I've added a warning when pulling/fetching with a configured cache `storage-service`. I've also expanded the documentation as per your input. Is this clear enough now?

> Beyond this, logging about pulling and pushing artifacts should be enhanced.

`bst artifact pull` will not pull anything if the specified artifact is already in the local cache or the cache storage-service. However, if the artifact is not cached yet, it will actually pull the blobs from the artifact server to the local cache. The reason for this is that CAS doesn't support transfers between two servers without going through the local client. Even though this is more of a side-effect of the protocol limitation, this means that one side of pulling/pushing is always localhost and thus, I'm not sure whether we even should adjust the log messages.

Reply:
> I've added a warning when pulling/fetching with a configured cache storage-service. I've also expanded the documentation as per your input. Is this clear enough now?

The documentation is much improved thanks for that :)

> The reason for this is that CAS doesn't support transfers between two servers without going through the local client.

I guess I was overly optimistic to think this might not be the case, in which case I agree we don't need to change logging.